### PR TITLE
Maple Mini Support for the K3NG

### PR DESCRIPTION
**Pull request by kc2uez**
Added STM32duino based Maple Mini support to the K3NG Keyer. I was having issues with the console, but it wasn't the code, it was the STM32duino having issues with the printing when F( is used. This was fixed yesterday #231 and now CLI works, LCD works, winkey emu, tonight I will test decoding and other functions.
I made a little video of me QLFing: https://www.youtube.com/watch?v=SBhDNS_c-Eo 
The maple mini is loaded with the maple-mini bootloader 2.0. and the IDE uses the latest STM32duino. 120kb flash and 20kb ram almost the same size of the arduino nano, The main changes were on the function prototyping. I kept everything in between #if define to keep it safe for other devices.  Ethernet broke the compilation proccess, so I took the comment out. It still compiles on the mega.

### CODE REVIEW

File: k3ng_keyer/k3ng_keyer.ino [hunk 1]
```diff
@@ -606,9 +606,38 @@ Recent Update History
   #define tone toneDUE
   #define noTone noToneDUE
 #elif defined(ARDUINO_MAPLE_MINI)
-  #include <SPI.h>
-  #include <Wire.h>
+//  #include <SPI.h>
+//  #include <Wire.h>
   #include <EEPROM.h>
+  void initialize_pins(void);
```
Comment (k3ng) [inline k3ng_keyer/k3ng_keyer.ino]: What IDE are you using to compile?  You get an error if you don't have these declarations?

File: k3ng_keyer/keyer_debug.h [hunk 2]
```diff
@@ -42,4 +42,4 @@
 // #define DEBUG_FORCE_RESET
 // #define DEBUG_WINKEY_DISABLE_LEAD_IN_TIME_SETTING
 
-// #define OPTION_WINKEY_IGNORE_FIRST_STATUS_REQUEST
+// #define OPTION_WINKEY_IGNORE_FIRST_STATUS_REQUEST
```
Comment (k3ng) [inline k3ng_keyer/keyer_debug.h]: This is specific to your configuration.

File: k3ng_keyer/keyer_features_and_options.h [hunk 3]
```diff
@@ -4,7 +4,7 @@
 
 // #define FEATURE_COMMAND_BUTTONS
 // #define FEATURE_COMMAND_LINE_INTERFACE  // Command Line Interface functionality
-// #define FEATURE_MEMORIES             // on the Arduino Due, you must have FEATURE_EEPROM_E24C1024 and E24C1024 EEPROM hardware in order to compile this
+ #define FEATURE_MEMORIES             // on the Arduino Due, you must have FEATURE_EEPROM_E24C1024 and E24C1024 EEPROM hardware in order to compile this
```
Comment (k3ng) [inline k3ng_keyer/keyer_features_and_options.h]: This is specific to your configuration.

File: k3ng_keyer/keyer_pin_settings.h [hunk 4]
```diff
@@ -2,17 +2,17 @@
...
+#define tx_key_line_1 19       // (high = key down/tx on)
+#define tx_key_line_2 0
 #define tx_key_line_3 0
 #define tx_key_line_4 0
 #define tx_key_line_5 0
 #define tx_key_line_6 0
-#define sidetone_line 4         // connect a speaker for sidetone
-#define potentiometer A0        // Speed potentiometer (0 to 5 V) Use pot from 1k to 10k
-#define ptt_tx_1 0              // PTT ("push to talk") lines
+#define sidetone_line 16         // connect a speaker for sidetone
```
Comment (k3ng) [inline k3ng_keyer/keyer_pin_settings.h]: This is specific to your configuration.

File: k3ng_keyer/keyer_pin_settings.h [hunk 5]
```diff
@@ -22,8 +22,8 @@
 #define tx_key_dah 0            // if defined, goes active for dah (any transmitter) - customized with tx_key_dit_and_dah_pins_active_state and tx_key_dit_and_dah_pins_inactive_state
 
 #ifdef FEATURE_COMMAND_BUTTONS
-  #define analog_buttons_pin A1
-  #define command_mode_active_led 0
+  #define analog_buttons_pin 20
```
Comment (k3ng) [inline k3ng_keyer/keyer_pin_settings.h]: This is specific to your configuration.

File: k3ng_keyer/keyer_pin_settings.h [hunk 6]
```diff
@@ -76,15 +76,15 @@
 #endif //FEATURE_PTT_INTERLOCK
 
 #ifdef FEATURE_STRAIGHT_KEY
-  #define pin_straight_key 52
+  #define pin_straight_key 32
```
Comment (k3ng) [inline k3ng_keyer/keyer_pin_settings.h]: This is specific to your configuration.

File: k3ng_keyer/keyer_pin_settings.h [hunk 7]
```diff
@@ -76,15 +76,15 @@
...
 #endif //FEATURE_STRAIGHT_KEY
 
 #ifdef FEATURE_CW_DECODER
   #define cw_decoder_pin A3//A11 //A5 //A3  
   #ifdef OPTION_CW_DECODER_GOERTZEL_AUDIO_DETECTOR
     #define cw_decoder_audio_input_pin 0 // this must be an analog pin!
   #endif //OPTION_CW_DECODER_GOERTZEL_AUDIO_DETECTOR
-  #define cw_decoder_indicator 24
+  #define cw_decoder_indicator 33
```
Comment (k3ng) [inline k3ng_keyer/keyer_pin_settings.h]: This is specific to your configuration.

File: k3ng_keyer/keyer_settings.h [hunk 8]
```diff
@@ -17,7 +17,7 @@
 #define initial_ptt_lead_time_tx6 0         // PTT lead time in mS
 #define initial_ptt_tail_time_tx6 10         // PTT tail time in mS
 #define initial_qrss_dit_length 1        // QRSS dit length in seconds
-#define initial_pot_wpm_low_value 13     // Potentiometer WPM fully CCW
+#define initial_pot_wpm_low_value 10     // Potentiometer WPM fully CCW
```
Comment (k3ng) [inline k3ng_keyer/keyer_settings.h]: This is specific to your configuration.

File: k3ng_keyer/k3ng_keyer.ino (hunk 1 above)
Comment (kc2uez) [inline k3ng_keyer/k3ng_keyer.ino]: Hi, this is the ide from arduino.cc version 1.6.12. The board support is from the STM32duino from: https://github.com/rogerclarkmelbourne/Arduino_STM32
It adds support for the stm32f1×× and other arm family of mcu's ($3 maple mini clone is what I am using now). Without the function declarations you will get these errors: http://pastebin.com/1mu9gtN2
After the function declaration the code compiles. BTW Sorry for sending the pin declaration file, I only wanted to send the changes to the .inorganic.

Maybe moving the function declaration to a stm32duino.h may be better idea?

**Comment by k3ng** [comment]
I agree moving the function declarations to a separate header file would be
preferable.  I'll do that and push an update.

**Comment by k3ng** [comment]
Update pushed.  Please test the latest code.  Thanks!

**Pull request closed without merging**